### ISSUE TITLE
System.setProperty( DEFAULT_PROFILES_PROPERTY_NAME. "customprofile ) not affects the default profile

### ISSUE THREAD

**Issue by daniv** [spring-projects/spring-boot#9398]
hi. 

i don't know if the intention was to ommit the property DEFAULT_PROFILES_PROPERTY_NAME
defined in `org.springframework.core.env.AbstractEnvironment`
when we moved to SpringBoot we found that the default profile 'spring.profiles.default'
was not loaded in remains always "default" 
the root of the issue is that in SpringApplication  class method `protected void configureProfiles`
the environment it's started with `environment.getActiveProfiles()` only and not 
`environment.getDefaultProfiles()`  too.

maybe is the intention. 
the workaround is to implement an `EnvironmentPostProcessor`, and "**build**" the default profile.
however it was not recognized on `org.springframework.boot.context.config.ConfigFileApplicationListener` yet.

so we had to implement an `ApplicationListener` to do this little job 
i would like to know wether this "isuue" for us was a design intention or ... a defect

thanx

**Comment by philwebb** [comment]
This sounds like a bug on our side but I'm not sure that we can fix it by changing `SpringApplication.configureProfiles`. Do you have a sample application that you can share that shows exactly how the problem manifests itself?

**Comment by daniv** [comment]
hi phil. 
well, i cannot expose the app. however it is very simple to reproduce, System.setProperty( "spring.profiles.default", "web-driver" ) before launching SpringApplication,
then, it can obviously be seen that the after org.springframework.boot.SpringApplication#prepareEnvironment still default profile remains "default" unless you starting it with `environment.getDefaultProfiles()`  like it is done in `#configureProfiles()` by activating the de #getActiveProfiles()

as i said, it is no critical beacause it can be workarounded in many ways, but because we need the profile properties to be loaded at `org.springframework.boot.context.config.ConfigFileApplicationListener` it was a little tricky ( override listerners Order etc ). anyway, you're doing a great job, and personally i think SpringBoot is a master piece. i wish to see more examples for non-web applications........
we are using SpringBoot as the engine of our testing framework.

have a nice day

**Comment by wilkinsona** [comment]
I can't reproduce what I understand to be the problem here. When I run this application:

```java
package com.example;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

@SpringBootApplication
public class Gh9398Application {

	public static void main(String[] args) {
		System.setProperty("spring.profiles.default", "web-driver");
		System.out.println(SpringApplication.run(Gh9398Application.class, args).getBean(Foo.class));
	}

	@Component
	@Profile("web-driver")
	static class Foo {

	}

}
```

The following is logged at startup:

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v1.5.3.RELEASE)

2017-06-06 10:52:45.352  INFO 15659 --- [           main] com.example.Gh9398Application            : Starting Gh9398Application on aw-rmbp.home with PID 15659 (/Users/awilkinson/dev/workspaces/spring/spring-boot/1.5.x/gh-9398/target/classes started by awilkinson in /Users/awilkinson/dev/workspaces/spring/spring-boot/1.5.x/gh-9398)
2017-06-06 10:52:45.354  INFO 15659 --- [           main] com.example.Gh9398Application            : No active profile set, falling back to default profiles: web-driver
2017-06-06 10:52:45.394  INFO 15659 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@5c18298f: startup date [Tue Jun 06 10:52:45 BST 2017]; root of context hierarchy
2017-06-06 10:52:45.801  INFO 15659 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Registering beans for JMX exposure on startup
2017-06-06 10:52:45.812  INFO 15659 --- [           main] com.example.Gh9398Application            : Started Gh9398Application in 0.667 seconds (JVM running for 0.933)
com.example.Gh9398Application$Foo@1c742ed4
2017-06-06 10:52:45.813  INFO 15659 --- [       Thread-2] s.c.a.AnnotationConfigApplicationContext : Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@5c18298f: startup date [Tue Jun 06 10:52:45 BST 2017]; root of context hierarchy
2017-06-06 10:52:45.814  INFO 15659 --- [       Thread-2] o.s.j.e.a.AnnotationMBeanExporter        : Unregistering JMX-exposed beans on shutdown
```

As you can see, `web-driver` is being used as the default profile and the `Foo` bean was created in the context.

If I remove the setting of the system property then the application fails to start due to the missing `Foo` bean:

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v1.5.3.RELEASE)

2017-06-06 10:53:39.631  INFO 15673 --- [           main] com.example.Gh9398Application            : Starting Gh9398Application on aw-rmbp.home with PID 15673 (/Users/awilkinson/dev/workspaces/spring/spring-boot/1.5.x/gh-9398/target/classes started by awilkinson in /Users/awilkinson/dev/workspaces/spring/spring-boot/1.5.x/gh-9398)
2017-06-06 10:53:39.634  INFO 15673 --- [           main] com.example.Gh9398Application            : No active profile set, falling back to default profiles: default
2017-06-06 10:53:39.686  INFO 15673 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@5c18298f: startup date [Tue Jun 06 10:53:39 BST 2017]; root of context hierarchy
2017-06-06 10:53:40.102  INFO 15673 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Registering beans for JMX exposure on startup
2017-06-06 10:53:40.113  INFO 15673 --- [           main] com.example.Gh9398Application            : Started Gh9398Application in 0.696 seconds (JVM running for 1.02)
Exception in thread "main" 2017-06-06 10:53:40.114  INFO 15673 --- [       Thread-2] s.c.a.AnnotationConfigApplicationContext : Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@5c18298f: startup date [Tue Jun 06 10:53:39 BST 2017]; root of context hierarchy
org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.example.Gh9398Application$Foo' available
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:353)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:340)
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1093)
	at com.example.Gh9398Application.main(Gh9398Application.java:13)
2017-06-06 10:53:40.115  INFO 15673 --- [       Thread-2] o.s.j.e.a.AnnotationMBeanExporter        : Unregistering JMX-exposed beans on shutdown
```

@daniv Have I misunderstood the problem? To remove any confusion, can you please provide a small example that illustrates the problem that you're seeing.

**Comment by daniv** [comment]
hi 

i wrote a little program trying to reproduce the problem and i have the following conclusions.
when i defined `System.setProperty("spring.profiles.default", "Gh9398");` and ONLY that i got the same log as you have

s.boot.bazar.BazarShopApplication        : No active profile set, falling back to default profiles: Gh9398
ok, according the documentation if no active profile was set , the "default" will be activated, which is exactly what happended.

However, in my application properties i have the following definition 
**spring.profiles.active              : excel,web-driver**

now i get this log:
`INFO 5156 --- [           main] s.boot.bazar.BazarShopApplication        : The following profiles are active: excel,web-driver`
i understand that spring took my "ACTIVE" definition, however he discarded my "DEFAULT" definition
form the documentation i don't see that default will be discarded , but it  will became active if no other profiles were defined.

i was expecting the active profile list to be 
INFO 5156 --- [           main] s.boot.bazar.BazarShopApplication        : The following profiles are active: excel,web-driver,gh9398

**Comment by wilkinsona** [comment]
Thanks. This is working as designed. From [the documentation](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config-profile-specific-properties):

> The Environment has a set of default profiles (by default [default]) which are used if no active profiles are set

**Comment by daniv** [comment]
hi

as i said the documentation is very clear " (by default [default]) which are used if no active profiles are set"
but this is not why i raised this issue.
the documentation does not say that having also active profiles it will "DISCARD" or "NOT INCLUDE" those who were defined in `"spring.profiles.default"`
from you, i understand that this is the designed behaviuor....so maybe it should added to the documentation...

i could make a workaround by create a listener that just have the code environment.getDefaultProfiles()
and they magically included in the final active profile list.

anyway, thanx for clarify this point, and again, maybe you should update the documentation on active/additional/default profiles 

i also achieved it by extending SpringApplication and the method by adding the default profiles call

`	protected void configureProfiles( ConfigurableEnvironment environment, String[] args )
	{
		String[] activeProfiles = environment.getActiveProfiles();
		String[] defaultProfiles = environment.getDefaultProfiles();`

of course is a too big change just to change the behaviour of the  "configureProfiles" method

**Comment by wilkinsona** [comment]
The documentation describes the default profiles as being "used if no active profiles are set". You have set an active profile so the default profiles are not used. Sorry, but I'm not sure how that could be made any clearer. Perhaps you have a suggestion?

**Comment by daniv** [comment]
just to add a comment/paragraph/note to the documentation that also explains that by defined an active profile all the "default.profiles" will be ignored.
but again, i wanted to understand it by myself why suddenly the "default.profiles" were not merged in spring boot, and we got the conclusion that nowhere "environment.getDefaultProfiles()" was called.

now, of course we adapted the code to this specification, we wanted to be 100%  sure that this is the designed behaviour.  
thank u very much sr. for clarify this point
